aspnetcore/grpc/troubleshoot.md

```diff
@@ -0,0 +1,116 @@
+
+```csharp
+// This switch must be set before creating the HttpClient.
+AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
+
+var httpClient = new HttpClient();
+// The port number(5000) must match the port of the gRPC server.
+httpClient.BaseAddress = new Uri("http://localhost:5000");
+var client = GrpcClient.Create<Greeter.GreeterClient>(httpClient);
+```
+
+## Unable to start ASP.NET Core gRPC app on macOS
+
+Kestrel doesn't support HTTP/2 with TLS on macOS. The ASP.NET Core gRPC template and samples use TLS by default. You'll see the following error message when you attempt to start the gRPC server:
+
+> Unable to bind to https://localhost:5001 on the IPv4 loopback interface: 'HTTP/2 over TLS is not supported on macOS due to missing ALPN support.'.
+
+To work around this issue, configure Kestrel and the gRPC client to use HTTP/2 **without** TLS. You should only do this during development. Not using TLS will result in gRPC messages being sent without encryption.
+
+Kestrel must configure an HTTP/2 endpoint without TLS in *Program.cs*:
+
+```csharp
+public static IHostBuilder CreateHostBuilder(string[] args) =>
+    Host.CreateDefaultBuilder(args)
+        .ConfigureWebHostDefaults(webBuilder =>
+        {
+            webBuilder.ConfigureKestrel(options =>
+            {
+                // Setup a HTTP/2 endpoint without TLS.
+                options.ListenLocalhost(5000, o => o.Protocols = HttpProtocols.Http2);
+            });
+            webBuilder.UseStartup<Startup>();
+        });
+```
+
+The gRPC client must also be configured to not use TLS. For more information, see [Call insecure gRPC services with .NET Core client](#call-insecure-grpc-services-with-net-core-client).
+
+> [!WARNING]
+> HTTP/2 without TLS should only be used during app development. Production apps should always use transport security. For more information, see [Security considerations in gRPC for ASP.NET Core](xref:grpc/security#transport-security).
+
+## gRPC C# assets are not code generated from *\*.proto* files
+
+gRPC code generation of concrete clients and service base classes requires protobuf files and tooling to be referenced from a project. You must include:
+
+* *.proto* files you want to use in the `<Protobuf>` item group.
+* Package reference to the gRPC tooling package [Grpc.Tools](https://www.nuget.org/packages/Grpc.Tools/).
+
+For more information on generating gRPC C# assets, see <xref:grpc/basics>.
+
+By default a `<Protobuf>` reference will generate both a concrete client and a service base class. `GrpcServices` attribute can be used to limit C# asset generation to either the concrete client or service base class. Valid `GrpcServices` options are:
+
+* `Server`
+* `Client`
+* `Both` (the default)
+
+An ASP.NET Core web app hosting gRPC services only needs the service base class generated:
```

Comment:
Is it worth adding that imported files with message definitions only still need to be included. We don't automatically run code generation on imports.